app/models/user.rb

```diff
@@ -57,6 +58,35 @@ def self.to_csv
     end
   end
 
+  def self.from_omniauth(access_token)
+    data = access_token[:info]
+    user = User.find_by(email: data[:email])
+    unless user
+      password = Devise.friendly_token[0,20]
+      user = User.create(first_name: data[:first_name],
+      last_name: data[:last_name],
+      email: data[:email],
+      role: ADMIN,
+      password: password,
+      password_confirmation: password,
+      google_token: access_token[:credentials][:token],
+      google_refresh_token: access_token[:credentials][:refresh_token]
+      )
+    end
+    user
+  end
+
+  def self.new_with_session(params, session)
+    super.tap do |user|
+      if data = session["devise.google_data"] && session["devise.google_data"]["extra"]["raw_info"]
```

Comment:
mistake here, you dont save extra data coz it's too big for session
and so this condition will never work

Reply:
i think this code is failing.. pls add feature scenario to verify